### PR TITLE
Add Initial support for ContextNet Encoder and CTC Decoder

### PR DESCRIPTION
**Pull request by titu1994**
# Changelog
## Added
- Add ContextNetEncoder, ContextNetDecoderForCTC neural modules to ASR collection
- Add `stride_last` flag which allows `stride` and `repeat` flags to be used simultaneously. It will perform the strided convolution at the final Conv-BN-ReLU sub-block.
- Add `swish` as optional activation function
- Add `zero_infinity` flag to `CTCLoss`, default to False.
- Adds integration test for ContextNetEncoder and ContextNetDecoderForCTC

## Modified
- Update Squeeze and Excitation sub-module to support different context sizes, support different activation
  - Change default `se_reduction_ratio` to 8 instead of 16.
- `SpecAugment` now supports either an integer or floating point value for `time_width`.
  - If float is passed, adaptively uses it as percentage of current timesteps that should be cut.

**Note**: Currently, `examples/asr/contextnet.py` uses JasperDecoderForCTC instead of ContextNetDecoderForCTC. This will be updated in a future PR once full support is present.

### CODE REVIEW

File: nemo/collections/asr/contextnet.py [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+import nemo
+from .parts.jasper import JasperBlock, init_weights, jasper_activations
+from nemo.backends.pytorch.nm import TrainableNM
+from nemo.core.neural_types import *
+from nemo.utils.decorators import add_port_docs
+
+logging = nemo.logging
+
+
+class ContextNetEncoder(TrainableNM):
```
Comment (okuchaiev) [inline nemo/collections/asr/contextnet.py:17]: Should this inherit from JasperEncoder ?

File: nemo/collections/asr/parts/jasper.py [hunk 2]
```diff
@@ -179,21 +179,58 @@ def forward(self, x):
...
+        """
+        Squeeze-and-Excitation sub-module.
+
+        Args:
+            channels: Input number of channels.
+            reduction_ratio: Reduction ratio for "squeeze" layer.
+            context_window: Integer number of timesteps that the context
+                should be computed over, using stride 1 average pooling.
+                If value < 1, then global context is computed.
+            interpolation_mode: Interpolation mode of timestep dimension.
```
Comment (okuchaiev) [inline nemo/collections/asr/parts/jasper.py:194]: what are the possible values here?

File: nemo/collections/asr/contextnet.py [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+    def _prepare_for_deployment(self):
+        m_count = 0
+        for m in self.modules():
+            if type(m).__name__ == "MaskedConv1d":
+                m.use_mask = False
+                m_count += 1
+        logging.warning(f"Turned off {m_count} masked convolutions")
+
+    def __init__(
+        self,
```
Comment (okuchaiev) [inline nemo/collections/asr/contextnet.py:148]: for new Modules/Classes please let's start using Python 3 type hints

File: nemo/collections/asr/losses.py [hunk 4]
```diff
@@ -41,11 +41,11 @@ def output_ports(self):
         # return {"loss": NeuralType(None)}
         return {"loss": NeuralType(elements_type=LossType())}
 
-    def __init__(self, num_classes):
+    def __init__(self, num_classes, zero_infinity=False):
```
Comment (okuchaiev) [inline nemo/collections/asr/losses.py:44]: can you please add docstring to this class

File: nemo/collections/asr/parts/jasper.py [hunk 5]
```diff
@@ -179,21 +179,58 @@ def forward(self, x):
...
+                Used only if context window is > 1.
+            activation: Intermediate activation function used.
```
Comment (okuchaiev) [inline nemo/collections/asr/parts/jasper.py:196]: should this be string or func?

File: nemo/collections/asr/contextnet.py (hunk 1 above)
Comment (titu1994) [reply nemo/collections/asr/contextnet.py:17]: Good point

File: nemo/collections/asr/contextnet.py (hunk 3 above)
Comment (titu1994) [reply nemo/collections/asr/contextnet.py:148]: Sounds good.

File: nemo/collections/asr/losses.py (hunk 4 above)
Comment (titu1994) [reply nemo/collections/asr/losses.py:44]: Sounds good.

File: nemo/collections/asr/parts/jasper.py (hunk 2 above)
Comment (titu1994) [reply nemo/collections/asr/parts/jasper.py:194]: Actually, we dont expose this parameter for the user to override. The paper just uses `nearest`. I'll create another arg to allow to override this and update the docstring with possible values.

File: nemo/collections/asr/parts/jasper.py (hunk 5 above)
Comment (titu1994) [reply nemo/collections/asr/parts/jasper.py:196]: Function, ill note that in the docstring.

File: nemo/collections/asr/contextnet.py (hunk 1 above)
Comment (titu1994) [reply nemo/collections/asr/contextnet.py:17]: On second thought, it probably should not inherit JasperEncoder. While yes currently they share exactly same functionality, in the future they might not. In that case, the `__init__` call will instantiate multiple JasperBlocks before ContextNetEncoder starts to instantiate its own values.

While there is duplication for now, it is cleaner to separate the two modules

File: nemo/collections/asr/contextnet.py (hunk 3 above)
Comment (titu1994) [reply nemo/collections/asr/contextnet.py:152]: Fixed

File: nemo/collections/asr/losses.py (hunk 4 above)
Comment (titu1994) [reply nemo/collections/asr/losses.py:47]: Fixed

File: nemo/collections/asr/parts/jasper.py (hunk 2 above)
Comment (titu1994) [reply nemo/collections/asr/parts/jasper.py:199]: Fixed, added the `se_interpolation_mode` flag as a possible argument to `jasper`

File: examples/asr/contextnet.py [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+    )
+
+    # Inject the `kernel_size_factor` kwarg to the ContextNet config
+    # Skip the last layer  as that must be a pointwise kernel
+    for idx in range(len(contextnet_params["ContextNetEncoder"]["jasper"]) - 1):
+        contextnet_params["ContextNetEncoder"]["jasper"][idx]["kernel_size_factor"] = args.kernel_size_factor
+
+    # (ContextNet uses the Jasper baseline encoder and decoder)
+    encoder = nemo_asr.ContextNetEncoder(
+        feat_in=contextnet_params["AudioToMelSpectrogramPreprocessor"]["features"],
```
Comment (blisc) [inline examples/asr/contextnet.py:160]: Just a note that you can add this inside the yaml itself.
See https://confluence.atlassian.com/bitbucket/yaml-anchors-960154027.html

Comment (titu1994) [reply examples/asr/contextnet.py:160]: Thanks for the hint !

File: nemo/collections/asr/contextnet.py (hunk 1 above)
Comment (titu1994) [reply nemo/collections/asr/contextnet.py]: Fixed

**Pull request merged**